vaadin-renderer-flow-parent/vaadin-renderer-flow/src/main/resources/META-INF/resources/frontend/lit-renderer.ts

```diff
@@ -0,0 +1,112 @@
+import { render, html } from 'lit';
+
+type RenderRoot = HTMLElement & { __litRenderer?: Renderer; _$litPart$?: any };
+
+type ItemModel = { item: any; index: number };
+
+type Renderer = (
+  root: RenderRoot,
+  rendererOwner: HTMLElement,
+  model: ItemModel
+) => void;
+
+const _window = window as any;
+_window.Vaadin = _window.Vaadin || {};
+
+/**
+ * Assigns the component a renderer function which uses Lit to render
+ * the given template expression inside the render root element.
+ *
+ * @param component The host component to which the renderer runction is to be set
+ * @param rendererName The name of the renderer function
+ * @param templateExpression The content of the template literal passed to Lit for rendering.
+ * @param returnChannel A channel to the server.
+ * Calling it will end up invoking a handler in the server-side LitRenderer.
+ * @param clientCallables A list of function names that can be called from within the template literal.
+ * @param propertyNamespace LitRenderer-specific namespace for properties.
+ * Needed to avoid property name collisions between renderers.
+ */
+_window.Vaadin.setLitRenderer = (
+  component: HTMLElement,
+  rendererName: string,
+  templateExpression: string,
+  returnChannel: (name: string, itemKey: string, args: any[]) => void,
+  clientCallables: string[],
+  propertyNamespace: string
+) => {
+  // Dynamically created function that renders the templateExpression
+  // inside the given root element using Lit
+  const renderFunction = Function(`
+    "use strict";
+
+    const [render, html, returnChannel] = arguments;
+
+    return (root, {item, index}, itemKey) => {
+      ${clientCallables
+        .map((clientCallable) => {
+          // Map all the client-callables as inline functions so they can be accessed from the template literal
+          return `
+          const ${clientCallable} = (...args) => {
+            if (itemKey !== undefined) {
+              returnChannel('${clientCallable}', itemKey, args[0] instanceof Event ? [] : [...args]);
+            }
+          }`;
+        })
+        .join('')}
+
+      render(html\`${templateExpression}\`, root)
+    }
+  `)(render, html, returnChannel);
+
+  const renderer: Renderer = (root, _, { index, item }) => {
+    // Clean up the root element of any existing content
+    // (and Lit's _$litPart$ property) from other renderers
+    // TODO: Remove once https://github.com/vaadin/web-components/issues/2161 is done
+    if (root.__litRenderer !== renderer) {
+      root.innerHTML = '';
+      delete root._$litPart$;
```

Comment:
Removing `_$litPart$` may be no longer necessary here since https://github.com/vaadin/web-components/pull/2172 has been merged.

Reply:
👍

Reply:
`vaadin-virtual-list` does not clear the content rendered by the old renderer when the renderer changes (see https://github.com/vaadin/web-components/issues/2235), so this workaround is still necessary here.